fairscale/nn/misc/auto_wrap.py

```diff
@@ -0,0 +1,197 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+import contextlib
+from typing import Any, Callable, Dict, Generator, List, Optional, Tuple, Union
+
+import torch.nn as nn
+
+from fairscale.nn.data_parallel.fully_sharded_data_parallel import FullyShardedDataParallel
+from fairscale.nn.misc import checkpoint_wrapper
+
+FSDP_MODULE_EXCLUDE_WRAP = [nn.ModuleList, nn.ModuleDict]
+FSDP_MODULE_BLOCKLIST = [nn.MultiheadAttention]
+
+
+@contextlib.contextmanager
+def enable_wrap(module_blocklist: Optional[List] = None, **wrapper_kwargs: Any) -> Generator[None, None, None]:
+    """
+    Context manager to wrap modules in FullyShardedDataParallel.
+
+    Useful for when you'd like to apply the same parameters to all child modules
+    that you wrap. A particularly important use case is wrapping large layers so
+    that they get sharded (in-place) during initialization, to avoid running out of
+    system memory. Large layers can indicate that they should be sharded via
+    the ``wrap`` annotation and this context manager can provide the
+    exact configuration for these nested instances.
+
+    Usage::
+
+        with enable_wrap(**params):
+            # Wraps layer in FSDP by default if within context
+            self.l1 = wrap(torch.nn.Linear(5, 5))
+            # Wraps children modules by default based on min_num_params
+            self.l2 = auto_wrap(TransformerBlock(), min_num_params=1e8)
+
+    Args:
+        module_blocklist: List of additional Module Classes to not wrap when
+            using :func:`auto_wrap`. This is useful to exclude unsupported
+            modules when wrapping recursively.
+        **wrapper_kwargs: Configuration settings that will be passed to all ``wrap``
+            instances inside the context
+    """
+    with ConfigAutoWrap(module_blocklist, **wrapper_kwargs):
+        yield
+
+
+def wrap(
+    module: nn.Module,
+    cls: Callable = FullyShardedDataParallel,
```

Comment:
making this general is really good. I now see why this is in nn.misc now.

To solve the circular import issue,  maybe just put this file in nn.wrap or nn.wrap_context or something like that?

Another way is to let
cls = None

then, in the function:

if cls is None:
  import FSDP
  cls = FSDP

This has a bit more overhead. but probably not on critical path. I don't know any other more elegant solutions. :-(